src/framework/midi/internal/midideviceslistener.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "midideviceslistener.h"
+
+#include <memory>
+
+using namespace mu::midi;
+
+MidiDevicesListener::MidiDevicesListener()
+{
+    m_devicesUpdateThread = std::make_shared<std::thread>(updateDevices, this);
+}
+
+MidiDevicesListener::~MidiDevicesListener()
+{
+    if (!m_devicesUpdateThread) {
+        return;
+    }
+
+    m_devicesUpdateThread->join();
+    m_devicesUpdateThread = nullptr;
+}
+
+void MidiDevicesListener::reg(const ActualDevicesCallback& callback)
+{
+    m_actualDevicesCallback = callback;
+}
+
+mu::async::Notification MidiDevicesListener::devicesChanged() const
+{
+    return m_devicesChanged;
+}
+
+void MidiDevicesListener::updateDevices(MidiDevicesListener* self)
+{
+    self->doUpdateDevices();
+}
+
+void MidiDevicesListener::doUpdateDevices()
+{
+    while (true) {
```

Comment:
We need a flag like `m_running` (atomic), which must be set in false in the destructor, before the join, otherwise the join will never happen.

Reply:
Should `m_actualDevicesCallback` also be atomic? (Or I could just prohibit to set the callback while already running.)

Reply:
m_actualDevicesCallback not posible to be atomic, so we should add mutex for it, or better yes, just prohibit to set the callback while already running.